otherlibs/systhreads/decs.c

```diff
@@ -0,0 +1,899 @@
+#define CAML_INTERNALS
+
+#include "caml/alloc.h"
+#include "caml/domain_state.h"
+#include "caml/platform.h"
+#include "caml/custom.h"
+#include "caml/memory.h"
+#include "caml/fail.h"
+#include "caml/alloc.h"
+#include "caml/startup.h"
+#include "caml/fiber.h"
+#include "caml/callback.h"
+#include "caml/weak.h"
+#include "caml/finalise.h"
+#include "caml/domain.h"
+#include "caml/printexc.h"
+#include "caml/backtrace.h"
+#include <pthread.h>
+#include <signal.h>
+#include <caml/signals.h>
+
+/* ML value for a thread descriptor */
+
+struct caml_dec_descr {
+  value id;
+  value clos;
+  value terminated;
+};
+
+#define Decstatus_val(v) (* ((dec_event *) Data_custom_val(v)))
+#define Ident(v) (((struct caml_dec_descr *)(v))->id)
+#define Clos(v) (((struct caml_dec_descr *)(v))->clos)
+#define Terminated(v) (((struct caml_dec_descr *)(v))->terminated)
+
+/* Event structure */
+
+typedef struct event_struct {
+  caml_plat_mutex lock;         /* to protect contents */
+  int status;                   /* 0 = not triggered, 1 = triggered */
+  caml_plat_cond triggered;     /* signaled when triggered */
+} * dec_event;
+
+static value caml_decstatus_new (value unit);
+static int dec_event_create(dec_event * res);
+static void caml_decstatus_terminate (value wrapper);
+
+/* main runtime lock */
+
+struct caml_dec_main_lock {
+  caml_plat_mutex lock;
+  caml_plat_cond free;
+  atomic_uintnat busy;
+  atomic_uintnat waiters;
+};
+
+#ifdef NATIVE_CODE
+extern struct longjmp_buffer caml_termination_jmpbuf;
+extern void (*caml_termination_hook)(void);
+#endif
+
+/* dec structure holding runtime state */
+struct caml_dec_struct {
+
+  value descr;
+
+  struct caml_dec_struct * next;
+  struct caml_dec_struct * prev;
+
+  int domain_id;
+
+  struct stack_info* current_stack;
+  struct c_stack_link* c_stack;
+  struct caml__roots_block *local_roots;
+  struct longjmp_buffer *exit_buf;
+  int backtrace_pos;
+  code_t * backtrace_buffer;
+  caml_root backtrace_last_exn;
+  value * gc_regs_buckets;
+  value ** gc_regs_slot;
+  void * exn_handler;
+
+  #ifndef NATIVE_CODE
+  intnat trap_sp_off;
+  intnat trap_barrier_off;
+  struct caml_exception_context* external_raise;
+  #endif
+
+};
+
+typedef struct caml_dec_struct* caml_dec_t;
+
+/* overall table for decs accross domains */
+struct caml_dec_table {
+  caml_dec_t all_decs;
+  caml_dec_t current_dec;
+  pthread_key_t dec_key;
+  struct caml_dec_main_lock dec_lock;
+};
+
+/* dec_table instance, up to Max_domains */
+static struct caml_dec_table dec_table[Max_domains];
```

Comment:
Is it better to make this global state part of the per-domain `Caml_state`? Is there a reason for it to be here?

Reply:
But putting it in Caml_state would require us to have all the underlying definitions related to systhreads be a part of the runtime, if I'm not mistaking?
systhreads sits currently outside of the runtime